### ISSUE TITLE
Upsert does not work with edges and search-expression { _from: ..., _to: ...}

### ISSUE THREAD

**Issue by lewisdiamond** [arangodb/arangodb#1442]
If you try to do an upsert statement where your search-expression is an example document using _from and _to, the OLD document will never be found and an insert statement will always be generated, creating duplicates.

Tested in 2.6.4

**Comment by jsteemann** [comment]
Can you provide an example for this?
I am asking because I tried the following and it worked for me in 2.6:
```js
/* vertex collection with 2 vertices v1 and v2 */
db._drop("verty"); 
db._create("verty");
db.verty.save({ _key: "v1" }); 
db.verty.save({ _key: "v2" });  

/* edge collection with 2 edges */
db._drop("edgy"); 
db._createEdgeCollection("edgy"); 
db.edgy.save("verty/v1", "verty/v2", { what: "v1->v2" }); 
db.edgy.save("verty/v2", "verty/v1", { what: "v2->v1" }); 

/* output number of edges before the upsert query (should be 2) */
print("count before upsert: ", db.edgy.count()); 

/* run upsert query */
var query = "UPSERT { _from: 'verty/v1', _to: 'verty/v2' } " +
  "INSERT { _from: 'verty/v1', _to: 'verty/v2', what: 'v1->v2', isNew: true } " +
  "UPDATE { isNew: false } IN edgy RETURN { OLD: OLD, NEW: NEW }"; 
print("upsert query result: ", db._query(query).toArray()); 

/* output number of edges after the upsert query (should be 2) */
print("count after upsert: ", db.edgy.count());

/* print contents of edge collection. this shouldn't return an edge with 
   attribute value isNew: true */
print("edges after upsert: ", db.edgy.toArray());
```
The above upserts an edge based on _from and _to, and as the edge already exists it will go the *update* path.

**Comment by CoDEmanX** [comment]
Works for me too, one edge has `isNew: false`, the other doesn't have this attribute at all (v2.6.4 and 2.7.0-devel).

@jsteemann: one issue though - `RETURN {OLD, NEW}` does not work, seems the special variables do not support the object literal simplification.

**Comment by lewisdiamond** [comment]
@jsteemann Your example works and it's essentially the same as what I'm trying.

```
upsert { _from: "user/1", _to: "tag/ccf6184efe5e38da61ef2e9a30d1c1ea"} 
insert { _from: "user/1", _to: "tag/ccf6184efe5e38da61ef2e9a30d1c1ea", weight: 1} 
update {weight: OLD.weight+1}
in test
RETURN { old: OLD, new: NEW }
```

My statement __works__ on a newly created edge collection. It does __not work__ on an edge collection created in earlier version(s) (2.5 maybe?). So there seem to be a problem with the upsert on an upgraded edge collection from an undefined previous version.

**Comment by jsteemann** [comment]
I don't see a good reason for upserts having problems with collections from 2.5 or before. 
How many results does `db.test.inEdges("tag/ccf6184efe5e38da61ef2e9a30d1c1ea");` produce (with `test` being the edge collection from the older version)?

**Comment by lewisdiamond** [comment]
I truncated the edge collection in order to provide a list of documents in the edge collection and the issue is gone now.

I was able to reproduce with the following steps:

```
arangosh [_system]> db._drop("vertX")
arangosh [_system]> db._drop("vertY")
arangosh [_system]> db._drop("edge")
arangosh [_system]> db._create("vertX")
[ArangoCollection 1258062865882, "vertX" (type document, status loaded)]

arangosh [_system]> db._create("vertY")                                                                                                                                                                            
[ArangoCollection 1258062996954, "vertY" (type document, status loaded)]

arangosh [_system]> db._createEdgeCollection("edge")
[ArangoCollection 1258063586778, "edge" (type edge, status loaded)]

arangosh [_system]> db.vertX.save({test:1})
{ 
  "_id" : "vertX/1258065749466", 
  "_rev" : "1258065749466", 
  "_key" : "1258065749466" 
}

arangosh [_system]> db.vertY.save({test:2})
{ 
  "_id" : "vertY/1258066077146", 
  "_rev" : "1258066077146", 
  "_key" : "1258066077146" 
}
arangosh [_system]> db.edge.save("vertX/1258065749466", "vertX/1258065749466",{test:1})
{ 
  "_id" : "edge/1258073744858", 
  "_rev" : "1258073744858", 
  "_key" : "1258073744858" 
}

arangosh [_system]> db.edge.save("vertX/1258065749466", "vertY/1258066077146",{test:2})                                                                                                                           
{ 
  "_id" : "edge/1258075776474", 
  "_rev" : "1258075776474", 
  "_key" : "1258075776474" 
}
```

AQL:

```
upsert { _from: "vertX/1258065749466", _to: "vertY/1258066077146"} 
insert { _from: "vertX/1258065749466", _to: "vertY/1258066077146", test: 2} 
update {test: OLD.test+1}
in edge
RETURN { old: OLD, new: NEW }

[
  {
    "old": null,
    "new": {
      "test": 2,
      "_id": "edge/1258080626138",
      "_rev": "1258080626138",
      "_key": "1258080626138",
      "_from": "vertX/1258065749466",
      "_to": "vertY/1258066077146"
    }
  }
]

```
__TLDR:__ It seems like upsert gets upset (pun intended) when the edge links to different document types.

**Comment by jsteemann** [comment]
The code you posted recently revealed the problem cause: the problem occurs in 2.6 when running a query with 
* LIMIT 1 (UPSERT internally uses LIMIT 1) on an edge index 
* with both `_from` and `_to` in the FILTER condition (the UPSERT search value is a filter condition internally)
* and multiple edges matching the search criteria specified for `_from` 
* and the first edge returned from the index not being the one containing the search criteria specified for `_to`

I found an existing fix for this issue that had been added to devel already, but unfortunately not for 2.6. 
So I pushed the fix into 2.6 as well:
https://github.com/arangodb/arangodb/commit/7e614ce0c16bc723e4e2086c92cb97ef6926e9be

**Comment by lewisdiamond** [comment]
I built branch 2.6 and it seems to be working now. Thanks
